### PR TITLE
Update PHP-Scoper to v0.16.2 

### LINKED ISSUE #10407: Update PHP-Scoper to v0.16.2
## Task Description

The current version of PHP-Scoper is 0.16.2.

In [version 0.16.0](https://github.com/humbug/php-scoper/releases/tag/0.16.0), there were lots of improvements to the config, making it a lot simpler for us to configure PHP-Scoper in a WP plugin context.

It means we can get rid of all the manual `str_replace` calls in the scoper config thanks to `expose-global-{constants, classes, functions}` and `expose-{namespaces, constants, classes, functions}`

We could even automate this by setting the config based on the WP stubs, see https://github.com/humbug/php-scoper/issues/303#issuecomment-1021388471

Though we could split that up in its own ticket.

### Implementation Brief

* Install latest version of PHP-Scoper via Composer
* Remove all the `str_replace` calls in the PHP-Scoper config
* Install [`sniccwp/php-scoper-wordpress-excludes`](https://packagist.org/packages/sniccwp/php-scoper-wordpress-excludes) as a dev dependency and configure the `exclude-classes`, `exclude-functions`, and `exclude-constants` options accordingly

### PR DESCRIPTION
**Pull request by timarney**
## Context

Updates PHP-Scoper to 0.16.2 which allows us to clean up the config file

## Summary
Updates the PHP-Scoper version + config

- Removes "str_replace" calls in the config

## Relevant Technical Choices

- Pulls in `sniccwp/php-scoper-wordpress-excludes` which provides .json files for WP functions, classes to exclude

## User-facing changes

N/A

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### Does this PR have a security-related impact?

No outside of introduces a new package `sniccwp/php-scoper-wordpress-excludes`

### Does this PR change what data or activity we track or use?

No

### Does this PR have a legal-related impact?

No new package is MIT License

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #10407

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -24,12 +24,14 @@
     "automattic/vipwpcs": "^2.3",
     "dealerdirect/phpcodesniffer-composer-installer": "^0.7",
     "dg/bypass-finals": "1.1.3",
+    "php-stubs/wordpress-stubs": "^5.9",
```
Comment (swissspidy) [inline composer.json]: This is a transitive dependency of both `szepeviktor/phpstan-wordpress` and `sniccwp/php-scoper-wordpress-excludes`, so we shouldn't need to add it here explicitly

File: scoper.inc.php [hunk 2]
```diff
@@ -10,11 +10,14 @@
 
 use Isolated\Symfony\Component\Finder\Finder;
 
+$wp_classes   = json_decode( file_get_contents( 'vendor/sniccwp/php-scoper-wordpress-excludes/generated/exclude-wordpress-classes.json' ) );
+$wp_functions = json_decode( file_get_contents( 'vendor/sniccwp/php-scoper-wordpress-excludes/generated/exclude-wordpress-functions.json' ) );
```
Comment (swissspidy) [inline scoper.inc.php]: I think we can add the constants here too, then this should be good to go!

Comment (timarney) [reply scoper.inc.php]: If we're looking to pull from that package -- I'm seeing an empty array at the moment.

Screen Shot 2022-02-08 at 6 52 38 AM

Comment (swissspidy) [reply scoper.inc.php]: Hmm that's unexpected. Sounds like a bug in that package.

@calvinalkan why is the constants file empty? I'd expect constants like `WP_CONTENT_DIR` and `WP_CONTENT_URL` to be in there.

Comment (timarney) [reply scoper.inc.php]: Once we have the answer for that and assuming that's a bug I'll set it up to array merge the rest of the constants

File: scoper.inc.php [hunk 3]
```diff
@@ -10,11 +10,27 @@
...
+ * Gets a list of WP globals to ignore.
+ *
+ * @param string $json_file the file name to pull the excludes list from.
+ * @return array[] Array of WP Classes, WP Functions, or WP Constants to exclude.
+ */
+function get_excludes( string $json_file ): array {
+	try {
+		$excludes_vendor_path = 'vendor/sniccwp/php-scoper-wordpress-excludes/generated/';
+		return json_decode( file_get_contents( $excludes_vendor_path . $json_file ) );
+	} catch ( \Exception $exception ) {
```
Comment (swissspidy) [inline scoper.inc.php]: Was there a lint warning or something about this?

The simple `json_decode( file_get_contents( 'vendor/sniccwp/php-scoper-wordpress-excludes/generated/exclude-wordpress-classes.json' ) );` statements you had originally seemed fine to me

Comment (timarney) [reply scoper.inc.php]: I can take it back out if you like ... was looking a little repetitive + getting long with the array merge (which is juts pulling in an empty array at the moment).

File: scoper.inc.php (hunk 2 above)
Comment (swissspidy) [reply scoper.inc.php]: Filed an issue at https://github.com/sniccowp/php-scoper-wordpress-excludes/issues/2

Once it's fixed we can do the array merge as you say in a follow-up PR.

File: scoper.inc.php (hunk 3 above)
Comment (swissspidy) [reply scoper.inc.php]: We can discuss it today but it didn't feel like a big issue at all.

**Comment by calvinalkan** [comment]
Hey, I just noticed today that we misspelled the package name. Since I released it on packagist about 4 days or so I still decided to fix it.

So the correct name is: `sniccowp/php-scoper-wordpress-excludes`. I'm sorry for the inconvenience.

File: scoper.inc.php (hunk 3 above)
Comment (timarney) [reply scoper.inc.php]: reverted to the get contents statements + updated to pull from the `sniccowp/php-scoper-wordpress-excludes` package name

**Review by swissspidy — APPROVED** [review]

File: scoper.inc.php [hunk 4]
```diff
@@ -10,11 +10,15 @@
 
 use Isolated\Symfony\Component\Finder\Finder;
 
+$wp_classes   = json_decode( file_get_contents( 'vendor/sniccowp/php-scoper-wordpress-excludes/generated/exclude-wordpress-classes.json' ) );
+$wp_functions = json_decode( file_get_contents( 'vendor/sniccowp/php-scoper-wordpress-excludes/generated/exclude-wordpress-functions.json' ) );
+$wp_constants = json_decode( file_get_contents( 'vendor/sniccowp/php-scoper-wordpress-excludes/generated/exclude-wordpress-constants.json' ) );
```
Comment (spacedmonkey) [inline scoper.inc.php:15]: Should we check if these files exists before trying to use them? Maybe create a little helper function. 

```
function read_file( $path ){
      if( !is_readable( $path ) ){
           return [];
      }
      return json_decode( file_get_contents( $path ) );
}
```

File: scoper.inc.php [hunk 5]
```diff
@@ -173,99 +177,34 @@
...
 
-	// See https://github.com/humbug/php-scoper#constants--classes--functions-from-the-global-namespace.
-	'whitelist-global-classes'   => false,
+	'exclude-functions' => $wp_functions,
 
-	// See https://github.com/humbug/php-scoper#constants--classes--functions-from-the-global-namespace.
-	'whitelist-global-functions' => false,
+	'exclude-constants' => array_merge(
+		$wp_constants,
+		[
```
Comment (spacedmonkey) [inline scoper.inc.php:195]: Why are these needed? Don't they come from `$wp_constants`. If they are a temporary, what is the plan to get rid of them? If temporary, a comment / todo to fix these hardcoded values at a later date.

Comment (swissspidy) [reply scoper.inc.php:195]: Please see discussion at https://github.com/GoogleForCreators/web-stories-wp/pull/10511#discussion_r801352022

Basically, this package does not yet provide a list of constants, but most likely will in a future version

File: scoper.inc.php (hunk 4 above)
Comment (timarney) [reply scoper.inc.php:15]: There was a small discussion related to a helper function here --- but not related to the existence of the files 
https://github.com/GoogleForCreators/web-stories-wp/pull/10511#discussion_r802115710

Comment (swissspidy) [reply scoper.inc.php:15]: We're including a specific version of a package that is known to include these files.

If a file does not exist we absolutely need to be informed. If this fails we need to know.

File: scoper.inc.php (hunk 5 above)
Comment (spacedmonkey) [reply scoper.inc.php:195]: That comment thread was hidden so didnt see this. My comment about adding a todo / comment is still valid.

Comment (timarney) [reply scoper.inc.php:195]: I'll add a comment ... there's an issue filed here -- it's currently an empty file.

https://github.com/sniccowp/php-scoper-wordpress-excludes/issues/2

File: scoper.inc.php (hunk 4 above)
Comment (timarney) [reply scoper.inc.php:15]: > If a file does not exist we absolutely need to be informed. If this fails we need to know.

1) To confirm we do want to add a helper?
2) If yes so we want to echo to the console or throw -- (how do we want to be informed?)

**Comment by spacedmonkey** [comment]
@swissspidy Little confused why this was merged, there were still outstanding questions. https://github.com/GoogleForCreators/web-stories-wp/pull/10511#discussion_r803960676

File: scoper.inc.php (hunk 4 above)
Comment (swissspidy) [reply scoper.inc.php:15]: No helper needed.

I don't see a need for action here.

**Comment by swissspidy** [comment]
I simply missed that comment. Resolved now.

I don't see a need for further action here.

**Pull request merged**